keps/sig-auth/3130-kms-observability/README.md

```diff
@@ -0,0 +1,259 @@
+	Cipher               []byte   `protobuf:"bytes,2,opt,name=cipher,proto3" json:"cipher,omitempty"`
+	XXX_NoUnkeyedLiteral struct{} `json:"-"`
+	XXX_unrecognized     []byte   `json:"-"`
+	XXX_sizecache        int32    `json:"-"`
+}
+```
+
+The UID generated in the kube-apiserver will be used:
+
+1. For logging in the kube-apiserver. All envelope operations to the kms-plugin will be logged with the corresponding UID.
+   1. The UID will be logged using a wrapper in the kube-apiserver to ensure that the UID is logged in the same format and is always logged.
+   2. In addition to the UID, the kube-apiserver will also log non-sensitive metadata such as name, namespace and GroupVersionResource of the object that triggered the envelope operation.
+2. Sent to the kms-plugin as part of the `EncryptRequest` and `DecryptRequest` structs.
+
+### Test Plan
+
+Unit tests covering:
+
+1. Generation of UID for each envelope operation
+
+Integration test covering:
+
+1. Logging of UID in kube-apiserver
+2. UID in the `EncryptRequest` and `DecryptRequest`
+3. UID set to nil in the `EncryptRequest` and `DecryptRequest` when the feature gate is disabled
+   1. Confirm this results in byte equivalent data on the wire when compared to a 1.23 API server.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial unit and integration tests completed and enabled
+
+#### Beta
+
+- Gather feedback from providers using the feature
+- Any known bugs fixed
+
+#### GA
+
+- This is part of the KMS reference implementation
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+<!--
+Pick one of these and delete the rest.
+-->
+
+- Feature gate
+  - Feature gate name: `KMSUID`
+  - Components depending on the feature gate:
+    - kube-apiserver
+
+```go
+FeatureSpec{
+	Default: false,
+	LockToDefault: false,
+	PreRelease: featuregate.Alpha,
+}
+```
+
+###### Does enabling the feature change any default behavior?
+
+UID sent as part of the envelope operation is a change in the default behavior. This is backwards compatible.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, via the `KMSUID` feature gate. Disabling this gate will cause the API server to not send the UID as part of `Encrypt` or `Decrypt` envelope operation.
+
+### Monitoring Requirements
+
+###### How can someone using this feature know that it is working for their instance?
+
+- [x] Other (treat as last resort)
+  - Details: Logs in kube-apiserver, kms-plugin and KMS will be logged with the corresponding UID.
+
+###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?
+
+There should be no impact on the SLO with this change.
+
+###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?
+
+- [x] Other (treat as last resort)
```

Comment:
I suggest updating this with a way to inspect the response code for secret queries to check for higher error rates on upgrade and elevated rates in general.  It doesn't block alpha PRR review, but will come around again for beta.